include/lp_hash_table.h

```diff
@@ -0,0 +1,100 @@
+#ifndef DIL_LP_HASH_TABLE_H
+#define DIL_LP_HASH_TABLE_H
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <assert.h>
+
+template <typename T>
+struct lp_hash_table {
+private:
+    const char **keys;
+    T *values;
+    int buckets;
+
+    unsigned int hash(const char *key) {
+        // TODO: implement a better hashing method
+        return ((uintptr_t)key % buckets);
+    }
+
+public:
+
+    /*
+     * Constructor. Assumes that `elems > 0`
+     */
+    lp_hash_table(int elems) {
+        assert(elems > 0);
+
+        // TODO: maybe find a way to optimize this
+        buckets = elems;
+
+        // TODO: use our own allocator when it is ready.
+        keys = (const char **)malloc(buckets * sizeof(const char *));
+        values = (T *)malloc(buckets * sizeof(T));
+
+        for (int i = 0; i <= buckets; ++i) {
+            // We need to initialize the pointer array because the insertion loop will check on it's indices
+            keys[i] = nullptr;
+        }
+    };
+
+    /*
+     * @return - 0 means everything went ok, -1 means there was an error
+     */
+    int insert(const char *key, T value) {
+
+        int index = this->hash(key);
+        int i = index;
+
+        while(this->keys[i] != nullptr) {
```

Comment:
Notice that what you're doing here is essentially what find does. I think it pays off to think of a way to not duplicate this code.

Reply:
I have this underway already but I wanted to push as fast as possible because of the workspace structure I chose. That way if anyone else wants to contribute code they could have a starting point.

Reply:
Of course, either way, it's quite good this way!